hamp_radar/writezarr.py

```diff
@@ -0,0 +1,83 @@
+import xarray as xr
+from pathlib import Path
+from typing import Optional
+
+
+def ideal_zarrchunks(ds: xr.Dataset):
+    """returns ideal size of chunks for variables with these dimensions
+    so that chuunks are about 1MB. Addiontally cocx dimension (size=2)
+    is split into 2 seperate chunks and range dimension (size=512)
+    is sub-divided into 16 smaller chunks (size=32)"""
+    import warnings
+
+    frame = ds.sizes["frame"]
+    if frame > 2621440:
+        warnings.warn(
+            "Warning: number of frames is very large, consider changing ideal chunkshapes",
+            UserWarning,
+        )
+    return {
+        (): (),
+        ("frame",): (frame,),
+        ("frame", "cocx"): (frame, 1),
+        ("frame", "range", "cocx"): (8192, 32, 1),
+        ("frame", "range", "cocx", "fft", "iq"): (512, 32, 1, 256, 2),
+    }
+
+
+def zarr_chunks_encoding(ds: xr.Dataset):
+    ideal_chunks = ideal_zarrchunks(ds)
+    encoding = {
+        x: {"chunks": ideal_chunks[ds[x].dims]}
+        for x in list(ds.coords) + list(ds.data_vars)
+        if ds[x].chunks
+    }
+    return encoding
+
+
+def merge_encodings(e1: dict, e2: dict):
+    for x in set(e1.keys()) & set(e2.keys()):
+        e1[x].update(e2[x])
+    for x in set(e2.keys()) - set(e1.keys()):
+        e1[x] = e2[x]
+    return e1
+
+
+def zarr_encoding(ds: xr.Dataset, encoding: Optional[dict] = None):
+    if encoding is None:
+        return zarr_chunks_encoding(ds)
+    else:
+        return merge_encodings(encoding, zarr_chunks_encoding(ds))
+
+
+def prepare_daskarrays_for_zarrchunks(ds: xr.Dataset, encoding: dict):
+    def safe_chunks(current_chunks, desired_chunks):
+        return [max(i, c) for i, c in zip(current_chunks, desired_chunks)]
+
+    for x in list(ds.coords) + list(ds.data_vars):
+        if ds[x].chunks:
+            current_chunks = (c[0] for c in ds[x].chunks)
+            desired_chunks = encoding[x]["chunks"]
+            ds[x] = ds[x].chunk(safe_chunks(current_chunks, desired_chunks))
+    return ds
+
+
+def good_chunksizes(ds: xr.Dataset, maxchunksize_mbytes: int):
+    import numpy as np
+
+    def is_good_chunks(ds):
+        for x in list(ds.coords) + list(ds.data_vars):
+            if ds[x].chunks:
+                chunksize = np.prod([c[0] for c in ds[x].chunks])
+                chunksize_mbytes = ds[x].values.itemsize * chunksize / 1024 / 1024
+                yield chunksize_mbytes < maxchunksize_mbytes
+            yield True
+
+    return all(list(is_good_chunks(ds)))
+
+
+def write_iqdataset(ds: xr.Dataset, filename: Path):
+    encoding = zarr_encoding(ds)
+    ds = prepare_daskarrays_for_zarrchunks(ds, encoding)
```

Comment:
At this point, it's probably a bit too late to find a "good" common chunking. If you want to read the data at a reasonable common chunking, you want to do this in the moment where you convert from memmapped numpy arrays to dask arrays. Once you have dask arrays, it's a bit too late to "remove" the initial chunking, and you might not gain that much anymore from optimizing it.

Reply:
Ah okay, I'll change this then and make the chunking when we write the datasets the desired chunking rather than the dictionary it is currently.